metascore/Users.mo

```diff
@@ -0,0 +1,215 @@
+            let accountId = nextAccountId;
+            nextAccountId += 1;
+            accountId;
+        };
+
+        // Returns the account linked to the given principal.
+        public func getAccountByPrincipal(principal : Principal) : ?MAccount.Account {
+            switch (principals.get(principal)) {
+                case (null) { null; };
+                case (? accountId)  {
+                    return accounts.get(accountId);
+                };
+            };
+        };
+
+        // Stores the given account.
+        public func putAccount(account : MAccount.Account) {
+            accounts.put(account.id, account);
+            switch (account.stoicAddress) {
+                case (null) {};
+                case (? p)  { principals.put(p, account.id); };
+            };
+            switch (account.plugAddress) {
+                case (null) {};
+                case (? p)  { principals.put(p, account.id); };
+            };
+        };
+
+        public func createAccount(player : MPlayer.Player) : MAccount.Account {
+            let account = {
+                alias         = null;
+                avatar        = null;
+                flavorText    = null;
+                id            = getNextAccountId();
+                plugAddress   = switch (player) {
+                    case (#plug(p))  { ?p;   };
+                    case (_)         { null; };
+                };
+                primaryWallet = player;
+                stoicAddress  = switch (player) {
+                    case (#stoic(p)) { ?p;   };
+                    case (_)         { null; }; 
+                };
+            };
+            putAccount(account);
+            account;
+        };
+
+        // Gets the account, or creates one if not found, returns whether the account was created.
+        public func ensureAccount(player : MPlayer.Player) : (MAccount.Account, Bool) {
+            let principal = MPlayer.unpack(player);
+            switch (principals.get(principal)) {
+                case (null) {
+                    (createAccount(player), true);
+                };
+                case (? accountId) {
+                    switch (accounts.get(accountId)) {
+                        case (? account) {
+                            (account, false);
+                        };
+                        case (null) {
+                            (createAccount(player), true);
+                        };
+                    };
+                };
+            };
+        };
+
+        public func canBeLinked(
+            player    : MPlayer.Player,
+            newPlayer : MPlayer.Player,
+        ) : Result.Result<MAccount.Account, Text> {
+            // Check whether the other player is not already linked.
+            switch (getAccountByPrincipal(MPlayer.unpack(newPlayer))) {
+                case (? _)  { return #err("other principal already linked"); };
+                case (null) {};
+            };
```

Comment:
As we discussed, this should not be an error. When performing a link, both wallets are actually expected to already have an account associated with them. Frontend calls `ensureAccount` whenever someone connects a wallet. Instead, we need to confirm that both accounts linked to the principal pair are only linked to one (the correct) principal.